### ISSUE TITLE
EXWM windows lose focus when switching persps with persp-mode

### ISSUE THREAD

**Issue by LemonBreezes** [emacs-exwm/exwm#18]
I'm having a lot of difficulty getting persp-mode working together with EXWM. For some reason, if I switch to a Persp containing Discord, Discord does not gain focus unless I was previously in a persp containing an EXWM buffer.

**Comment by minad** [comment]
Do you have such focus issues only with `persp-mode` and only with Discord, or also other programs? I don't use Discord but with Chromium I also regularly observe focus issues, when switching buffers with `switch-to-buffer`, `consult-buffer` or switching between tabs. I use `tab-bar-mode` instead of perspectives. After switching buffers, Chromium will not gain focus. I don't observe such a behavior with other programs, e.g., a simple Xterm or Thunderbird.

**Comment by LemonBreezes** [comment]
> Do you have such focus issues only with `persp-mode` and only with Discord, or also other programs? I don't use Discord but with Chromium I also regularly observe focus issues, when switching buffers with `switch-to-buffer`, `consult-buffer` or switching between tabs. I use `tab-bar-mode` instead of perspectives. After switching buffers, Chromium will not gain focus. I don't observe such a behavior with other programs, e.g., a simple Xterm or Thunderbird.

Yeah, I think this can happen when switching buffer, switching window, etc, according to an old workaround I wrote years ago (but have disabled). Lately I don't use many programs outside of Emacs but definitely it seems to happen in Discord and not in Firefox. 

I wrote a workaround that doesn't work as well as I would like it to because it interrupts repeat mode and creates a race condition where if I switch workspaces and start typing before the timers finish it will be interrupted
```
(defvar +exwm-refocus-application--message "")
(defvar +exwm-refocus-application--delays '(0.015 0.03))
(defvar +exwm-refocus-application--timer nil)
(defvar +exwm-refocus-application--last-time 0)
(defvar +exwm-refocus-application--last-state nil)

;;;###autoload
(defun +exwm-refocus-application (&rest _)
  "Refocus input for the currently selected EXWM buffer, if any."
  (when (and (derived-mode-p 'exwm-mode)
             (not (memq +exwm-refocus-application--timer
                        timer-list))
             (> (float-time) (+ +exwm-refocus-application--last-time
                                0.01
                                (cl-reduce #'+ +exwm-refocus-application--delays))))
    (run-at-time (nth 0 +exwm-refocus-application--delays)
                 nil #'+exwm-refocus-application--timer)))

(defun +exwm-refocus-application--timer ()
  (when (derived-mode-p 'exwm-mode)
    (setq +exwm-refocus-application--message (current-message))
    (let ((+exwm-refocus-application--last-state (bound-and-true-p evil-state)))
      (minibuffer-with-setup-hook
          #'+exwm-refocus-application-minibuffer-quit-timer
        (read-string "")))))

(defun +exwm-refocus-application-minibuffer-quit-timer ()
  (setq +exwm-refocus-application--timer
        (run-at-time (nth 1 +exwm-refocus-application--delays) nil
                     (lambda ()
                       (run-at-time
                        0.0 nil
                        (lambda ()
                          (when +exwm-refocus-application--message
                            (minibuffer-message +exwm-refocus-application--message))
                          (pcase +exwm-refocus-application--last-state
                            ('insert (exwm-evil-core-insert))
                            ('normal (exwm-evil-core-normal))
                            (_ nil))))
                       (when (minibufferp)
                         (setq +exwm-refocus-application--last-time (float-time))
                         (throw 'exit nil))))))
```

Then add `+exwm-refocus-application` to as many focus-changing hooks as possible. This is probably the most major EXWM issue I've noticed throughout the years.

**Comment by minad** [comment]
I currently use the workaround from https://github.com/ch11ng/exwm/pull/933:

~~~ elisp
    (advice-add #'exwm-layout--hide
                :after (lambda (id)
                         (with-current-buffer (exwm--id->buffer id)
                           (setq exwm--ewmh-state
                                 (delq xcb:Atom:_NET_WM_STATE_HIDDEN exwm--ewmh-state))
                           (exwm-layout--set-ewmh-state id)
                           (xcb:flush exwm--connection))))
~~~

This workaround completely defeats the `_NET_WM_STATE_HIDDEN` wm hint, which may have unwanted power consumption consequences.

**Comment by LemonBreezes** [comment]
> I currently use the workaround from [ch11ng/exwm#933](https://github.com/ch11ng/exwm/pull/933) and [ch11ng/exwm#759](https://github.com/ch11ng/exwm/issues/759).
> 
> ```emacs-lisp
>     (advice-add #'exwm-layout--hide
>                 :after (lambda (id)
>                          (with-current-buffer (exwm--id->buffer id)
>                            (setq exwm--ewmh-state
>                                  (delq xcb:Atom:_NET_WM_STATE_HIDDEN exwm--ewmh-state))
>                            (exwm-layout--set-ewmh-state id)
>                            (xcb:flush exwm--connection))))
> ```
> 
> This workaround completely defeats the `_NET_WM_STATE_HIDDEN` wm hint, which may have unwanted power consumption consequences.

Man. I've been using this for a few days and it's a game-changer. EXWM is not broken anymore...

**Comment by LemonBreezes** [comment]
> Have you tried setting `exwm-layout-auto-iconify` to nil? That may help.
> 
> This might be a race between updating input focus (which we defer and serialize with a lock) and processing other X events (which we serialize with a per-connection lock). Unfortunately, we're using two different locks here.
> 
> This is a shot in the dark, but could you test this patch?
> 
> ```diff
> diff --git a/exwm-input.el b/exwm-input.el
> index f1f035c..fd8ed51 100644
> --- a/exwm-input.el
> +++ b/exwm-input.el
> @@ -324,7 +324,7 @@ defun exwm-input--update-focus-commit
>    "Attempt to update the window focus.
>  If we're currently updating the window focus, re-schedule a focus update
>  attempt later."
> -  (if exwm-input--update-focus-lock
> +  (if (or exwm-input--update-focus-lock (< 0 (slot-value exwm--connection 'event-lock)))
>        (exwm-input--update-focus-defer)
>      (let ((exwm-input--update-focus-lock t))
>        (exwm-input--update-focus exwm-input--update-focus-window))))
> ```

Hi. I haven't tried your patch but I've been using the `exwm-layout--hide` and the issue has not reoccured since.
